sqlite3_opt_serialize.go

```diff
@@ -0,0 +1,68 @@
+// +build !sqlite_omit_deserialize
+
+package sqlite3
+
+/*
+#ifndef USE_LIBSQLITE3
+#include <sqlite3-binding.h>
+#else
+#include <sqlite3.h>
+#endif
+#include <stdlib.h>
+#include <stdint.h>
+*/
+import "C"
+
+import (
+	"fmt"
+	"unsafe"
+)
+
+// Serialize returns a byte slice that is a serialization of the database.
+// If the database fails to serialize, a nil slice will be returned.
+//
+// See https://www.sqlite.org/c3ref/serialize.html
+func (c *SQLiteConn) Serialize(schema string) []byte {
+	if schema == "" {
+		schema = "main"
+	}
+	var zSchema *C.char
+	zSchema = C.CString(schema)
+	defer C.free(unsafe.Pointer(zSchema))
+
+	var sz C.sqlite3_int64
+	ptr := C.sqlite3_serialize(c.db, zSchema, &sz, 0)
+	if ptr == nil {
+		return nil
+	}
+	defer C.sqlite3_free(unsafe.Pointer(ptr))
+	return C.GoBytes(unsafe.Pointer(ptr), C.int(sz))
+}
+
+// Deserialize causes the connection to disconnect from the current database
+// and then re-open as an in-memory database based on the contents of the
+// byte slice. If deserelization fails, error will contain the return code
+// of the underlying SQLite API call.
+//
+// When this function returns, the connection is referencing database
```

Comment:
This is a violation of the rules of cgo. https://pkg.go.dev/cmd/cgo#hdr-C_references_to_Go

> C code may not keep a copy of a Go pointer after the call returns.

Reply:
Thank you for pointing that out.

Reply:
Actually, why is it an issue? I presume you're talking about the `b` that is passed in -- but its lifetime is managed by the *caller* for this function. The caller will control when it goes out of scope. Or am I missing something? To be clear, I want to change this, the current implementation is clumsy, but want to be sure I understand the issue.

Do I first have to copy the Go slice to C memory space (or something like that?).

Reply:
Unless the caller explicitly does something to force it to remain in scope (e.g., `runtime.KeepAlive`) it can actually be garbage collected immediately. And even if they do (or you added `runtime.KeepAlive` here), it is still a violation as the Go runtime reserves the right to move stuff around in memory (right now it doesn't but may in the future).

I think you will need to first copy the Go `[]byte` into memory obtained via `sqlite3_malloc`. Then you can pass `SQLITE_DESERIALIZE_FREEONCLOSE` to to `sqlite3_deserialize`. I cannot think of any other way to make it truly safe.

Reply:
What do you suggest as the most efficient way to copy the memory from C to the Go slice? Use memcpy under the covers, something like? https://gist.github.com/3d0c/d8583aa8ac8fbc43585620a20a8b9ac9

Reply:
That comment in mistaken. I meant copying in the other direction.

Do you suggest using memcpy to copy data from the Go slice to the memory address allocated by `sqlite3_malloc`?

Reply:
As per https://github.com/mattn/go-sqlite3/pull/1089#discussion_r969670006

Reply:
Fixed in the code.